helix-view/src/editor.rs

```diff
@@ -271,6 +271,49 @@ impl Editor {
         Ok(())
     }
 
+    /// Refreshes the language server for a given document
+    pub fn refresh_language_server(&mut self, doc_id: DocumentId) -> Option<()> {
+        let doc = self.documents.get_mut(&doc_id)?;
+        doc.detect_language(Some(&self.theme), &self.syn_loader);
+        let language_server = Editor::get_lang_server(&mut self.language_servers, doc);
+
+        if let Some(language_server) = language_server {
+            let language_id = doc
+                .language()
+                .and_then(|s| s.split('.').last()) // source.rust
+                .map(ToOwned::to_owned)
+                .unwrap_or_default();
+
+            // TODO: this now races with on_init code if the init happens too quickly
+            tokio::spawn(language_server.text_document_did_open(
+                doc.url().unwrap(),
+                doc.version(),
+                doc.text(),
+                language_id,
+            ));
+
+            doc.set_language_server(Some(language_server));
+        }
```

Comment:
This whole `if` block is still duplicated from `open` and should be a part of `get_lang_server` (I'd name it something else)

Reply:
Would you consider it be better fitted under the `Document` struct as well?